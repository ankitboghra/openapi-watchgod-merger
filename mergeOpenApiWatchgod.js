/*
DEPOSIT
=> priority of responses(highest to lowest)
- openapiDeposits
- watchgodDeposits
*/

/*
WITHDRAW

=> priority of responses(highest to lowest)
- openapiExits
- openapiBurns
- watchgodExits
- watchgodBurns
*/

// interface is just for reference
/*
interface ITransaction {
    from?: String
    to?: String
    amount?: String
    userAddress?: String

    txType?: String
    txStatus?: String
    contractAddress?: String
    rootToken?: String

    network?: String

    tokenId?: String
    tokenType: String
    isPoS?: Boolean

    txHash?: String
    prevBurnHash?: String
    oldHash?: String
    newHash?: String

    data?: String
    blockNumber?: Number

    timestamp?: Number // in seconds
}
*/

const {
    convertOpenapiToInterface,
    convertWatchgodToInterface,
    convertInterfaceToWatchgod,
} = require("./utils");
/**
 *
 * @param {*} openapiDepositResponse
 * @param {*} openapiWithdrawResponse
 * @param {*} openapiBurnResponse
 * @param {*} watchGodResponse
 */
function merger(
    openapiDepositResponse,
    openapiWithdrawResponse,
    openapiBurnResponse,
    watchGodResponse
) {
    const openapiDeposits = openapiDepositResponse.map((tx) =>
        convertOpenapiToInterface(tx)
    );

    const watchgodDeposits = watchGodResponse
        .filter((tx) => tx.txType === "deposit")
        .map((tx) => convertWatchgodToInterface(tx));

    const openapiExits = openapiWithdrawResponse.map((tx) =>
        convertOpenapiToInterface(tx)
    );

    const openapiBurns = openapiBurnResponse.map((tx) =>
        convertOpenapiToInterface(tx)
    );

    const watchgodExits = watchGodResponse
        .filter((tx) => tx.txType === "exit")
        .map((tx) => convertWatchgodToInterface(tx));

    const watchgodBurns = watchGodResponse
        .filter((tx) => tx.txType === "withdraw")
        .map((tx) => convertWatchgodToInterface(tx));

    return [
        ...mergeDeposits(openapiDeposits, watchgodDeposits),
        ...mergeWithdraws(openapiExits, openapiBurns, watchgodExits, watchgodBurns),
    ];
}

function mergeDeposits(openapiDeposits, watchgodDeposits) {
    const mergedDeposits = {};

    watchgodDeposits.forEach((tx) => {
        const txHash = tx.txHash;
        mergedDeposits[txHash] = tx;
        mergedDeposits[txHash].latestStatus = tx.txStatus;
        mergedDeposits[txHash].watchgodTxStatus = tx.txStatus;
        mergedDeposits[txHash].txSource = "watchgod_deposits";
    });

    openapiDeposits.forEach((tx) => {
        const txHash = tx.txHash;
        const watchgodTxStatus = mergedDeposits[txHash]
            ? mergedDeposits[txHash].txStatus
            : false;

        mergedDeposits[txHash] = tx;
        mergedDeposits[txHash].latestStatus = tx.txStatus;
        mergedDeposits[txHash].openapiTxStatus = tx.txStatus;

        if (watchgodTxStatus) {
            mergedDeposits[txHash].watchgodTxStatus = watchgodTxStatus;
        }

        mergedDeposits[txHash].txSource = "openapi";
    });

    return Object.values(mergedDeposits);
}

function mergeWithdraws(
    openapiExits,
    openapiBurns,
    watchgodExits,
    watchgodBurns
) {
    const mergedWithdraws = {};

    watchgodBurns.forEach((tx) => {
        const txHash = tx.txHash;

        mergedWithdraws[txHash] = tx;
        mergedWithdraws[txHash]._txSource = "watchgod_burns";
        mergedWithdraws[txHash]._burnTxHash = tx.txHash;
        mergedWithdraws[txHash]._watchgodBurnTxStatus = tx.txStatus;
        mergedWithdraws[txHash]._latestStatus = tx.txStatus;
    });

    openapiBurns.forEach((tx) => {
        const txHash = tx.txHash;
        const watchgodBurnTxStatus =
            mergedWithdraws[txHash] && mergedWithdraws[txHash]._watchgodBurnTxStatus
                ? mergedWithdraws[txHash]._watchgodBurnTxStatus
                : false;

        mergedWithdraws[txHash] = tx;
        mergedWithdraws[txHash]._txSource = "openapi_burns";
        mergedWithdraws[txHash]._burnTxHash = tx.txHash;
        mergedWithdraws[txHash]._openapiBurnTxStatus = tx.txStatus;
        mergedWithdraws[txHash]._latestStatus = tx.txStatus;

        if (watchgodBurnTxStatus) {
            mergedWithdraws[txHash]._watchgodBurnTxStatus = watchgodBurnTxStatus;
        }
    });

    watchgodExits.forEach((tx) => {
        const txHash = tx.txBurnHash;

        const watchgodBurnTxStatus =
            mergedWithdraws[txHash] && mergedWithdraws[txHash]._watchgodBurnTxStatus
                ? mergedWithdraws[txHash]._watchgodBurnTxStatus
                : false;

        const openapiBurnTxStatus =
            mergedWithdraws[txHash] && mergedWithdraws[txHash]._openapiBurnTxStatus
                ? mergedWithdraws[txHash]._openapiBurnTxStatus
                : false;

        mergedWithdraws[txHash] = tx;
        mergedWithdraws[txHash]._txSource = "watchgod_exits";
        mergedWithdraws[txHash]._burnTxHash = tx.txBurnHash;
        mergedWithdraws[txHash]._exitTxHash = tx.txHash;
        mergedWithdraws[txHash]._watchgodExitTxStatus = tx.txStatus;
        mergedWithdraws[txHash]._latestStatus = tx.txStatus;

        if (watchgodBurnTxStatus) {
            mergedWithdraws[txHash]._watchgodBurnTxStatus = watchgodBurnTxStatus;
        }

        if (openapiBurnTxStatus) {
            mergedWithdraws[txHash]._openapiBurnTxStatus = openapiBurnTxStatus;
        }
    });

    openapiExits.forEach((tx) => {
        const txHash = tx.txBurnHash;

        const watchgodBurnTxStatus =
            mergedWithdraws[txHash] && mergedWithdraws[txHash]._watchgodBurnTxStatus
                ? mergedWithdraws[txHash]._watchgodBurnTxStatus
                : false;

        const openapiBurnTxStatus =
            mergedWithdraws[txHash] && mergedWithdraws[txHash]._openapiBurnTxStatus
                ? mergedWithdraws[txHash]._openapiBurnTxStatus
                : false;

        const watchgodExitTxStatus =
            mergedWithdraws[txHash] && mergedWithdraws[txHash]._watchgodExitTxStatus
                ? mergedWithdraws[txHash]._watchgodExitTxStatus
                : false;

        mergedWithdraws[txHash] = tx;
        mergedWithdraws[txHash]._txSource = "openapi_exits";
        mergedWithdraws[txHash]._burnTxHash = tx.txBurnHash;
        mergedWithdraws[txHash]._exitTxHash = tx.txHash;
        mergedWithdraws[txHash]._openapiExitTxStatus = tx.txStatus;
        mergedWithdraws[txHash]._latestStatus = tx.txStatus;

        if (watchgodBurnTxStatus) {
            mergedWithdraws[txHash]._watchgodBurnTxStatus = watchgodBurnTxStatus;
        }

        if (openapiBurnTxStatus) {
            mergedWithdraws[txHash]._openapiBurnTxStatus = openapiBurnTxStatus;
        }

        if (watchgodExitTxStatus) {
            mergedWithdraws[txHash]._watchgodExitTxStatus = watchgodExitTxStatus;
        }
    });

    return Object.values(mergedWithdraws);
}

module.exports = merger;
