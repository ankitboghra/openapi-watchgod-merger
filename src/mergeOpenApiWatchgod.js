/*
DEPOSIT
=> priority of responses(highest to lowest)
- openapiDeposits
- watchgodDeposits
*/

/*
WITHDRAW

=> priority of responses(highest to lowest)
- openapiExits
- openapiBurns
- watchgodExits
- watchgodBurns
*/

// interface is just for reference
/*
interface ITransaction {
    from?: String
    to?: String
    amount?: String
    userAddress?: String

    txType?: String
    txStatus?: String
    contractAddress?: String
    rootToken?: String

    network?: String

    tokenId?: String
    tokenType: String
    isPoS?: Boolean

    txHash?: String
    prevBurnHash?: String
    oldHash?: String
    newHash?: String

    data?: String
    blockNumber?: Number

    timestamp?: Number // in seconds
}
*/

import {
    convertOpenapiToInterface,
    convertWatchgodToInterface,
    convertInterfaceToWatchgod,
} from "./utils";
import { TX_TYPE } from './constants'


import mergeDeposits from './mergeDeposits'
import mergeWithdraws from './mergeWithdraws'

/**
 *
 * @param {*} openapiDepositResponse
 * @param {*} openapiWithdrawResponse
 * @param {*} openapiBurnResponse
 * @param {*} watchGodResponse
 */
export default function merger(
    openapiDepositResponse,
    openapiWithdrawResponse,
    openapiBurnResponse,
    watchGodResponse
) {
    const openapiDeposits = openapiDepositResponse.map((tx) =>
        convertOpenapiToInterface(tx)
    );

    const watchgodDeposits = watchGodResponse
        .filter((tx) => tx.txType === TX_TYPE.DEPOSIT)
        .map((tx) => convertWatchgodToInterface(tx));

    const openapiExits = openapiWithdrawResponse.map((tx) =>
        convertOpenapiToInterface(tx)
    );

    const openapiBurns = openapiBurnResponse.map((tx) =>
        convertOpenapiToInterface(tx)
    );

    const watchgodExits = watchGodResponse
        .filter((tx) => tx.txType === TX_TYPE.EXIT)
        .map((tx) => convertWatchgodToInterface(tx));

    const watchgodConfirmExits = watchGodResponse
        .filter((tx) => tx.txType === TX_TYPE.CONFIRM_EXIT)
        .map((tx) => convertWatchgodToInterface(tx));

    const watchgodBurns = watchGodResponse
        .filter((tx) => tx.txType === TX_TYPE.WITHDRAW)
        .map((tx) => convertWatchgodToInterface(tx));

    return [
        ...mergeDeposits(openapiDeposits, watchgodDeposits),
        ...mergeWithdraws(openapiExits, openapiBurns, watchgodExits, watchgodConfirmExits, watchgodBurns),
    ];
}
