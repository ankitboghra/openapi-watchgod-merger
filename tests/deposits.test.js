import mergeOpenApiWatchgod from '../src/index.js'

const watchgodDeposit_1 = {
    hash: "0xtxHash_1",
    network: "main",
    amount: "100000000000000000000",
    prevBurnHash: null,
    rootToken: "0xrootToken_1",
    status: "confirmed",
    txType: "deposit",
    blockNumber: 13816531,
    data:
        "0xdata_1",
    from: "0xfrom_1",
    timestamp: 1639662471018,
    to: "0xto_1",
}

const watchgodDeposit_2 = {
    hash: "0xtxHash_2",
    network: "main",
    amount: "100000000000000000000",
    prevBurnHash: null,
    rootToken: "0xrootToken_2",
    status: "confirmed",
    txType: "deposit",
    blockNumber: 13816531,
    data:
        "0xdata_2",
    from: "0xfrom_2",
    timestamp: 1639662471018,
    to: "0xto_2",
}

const openApiDeposit_1 = {
    transactionStatus: "deposited",
    id: "id_1",
    rootToken: "0xrootToken_1",
    tokenType: "tokenType_1",
    amount: "100000000000000000000",
    tokenId: null,
    depositor: "0xdepositer_1",
    userAddress: "0xuserAddress_1",
    transactionHash: "0xtxHash_1",
    timestamp: "2021-12-16T13:47:43.000Z",
    rootTunnelAddress: "0xrootTunnelAddress_1",
    amounts: null,
    blockNumber: null,
    isPos: null,
    tokenIds: null,
}

const openApiDeposit_2 = {
    transactionStatus: "deposited",
    id: "id_2",
    rootToken: "0xrootToken_2",
    tokenType: "tokenType_2",
    amount: "100000000000000000000",
    tokenId: null,
    depositor: "0xdepositer_2",
    userAddress: "0xuserAddress_2",
    transactionHash: "0xtxHash_2",
    timestamp: "2021-12-16T13:47:43.000Z",
    rootTunnelAddress: "0xrootTunnelAddress_2",
    amounts: null,
    blockNumber: null,
    isPos: null,
    tokenIds: null,
}


test('watchgod deposit only', () => {
    mergeOpenApiWatchgod([], [], [], [watchgodDeposit_1]).forEach((tx) => {
        expect(tx).toMatchSnapshot()
        expect(tx).toHaveProperty('_latestStatus', 'confirmed')
        expect(tx).toHaveProperty('_watchgodTxStatus', 'confirmed')
        expect(tx).toHaveProperty('_txSource', 'watchgod_deposits')
    })
});

test('watchgod deposits only', () => {
    mergeOpenApiWatchgod([], [], [], [watchgodDeposit_1, watchgodDeposit_2]).forEach((tx) => {
        expect(tx).toMatchSnapshot()
    })
});

test('openapi deposit only', () => {
    mergeOpenApiWatchgod([openApiDeposit_1], [], [], []).forEach((tx) => {
        expect(tx).toMatchSnapshot()
    })
});

test('openapi deposits only', () => {
    mergeOpenApiWatchgod([openApiDeposit_1, openApiDeposit_2], [], [], []).forEach((tx) => {
        expect(tx).toMatchSnapshot()
    })
});

test('openapi and watchgod', () => {
    mergeOpenApiWatchgod([openApiDeposit_1], [], [], [watchgodDeposit_1]).forEach((tx) => {
        expect(tx).toMatchSnapshot()
    })
});

test('openapi and watchgod', () => {
    mergeOpenApiWatchgod([openApiDeposit_1, openApiDeposit_2], [], [], [watchgodDeposit_1, watchgodDeposit_2]).forEach((tx) => {
        expect(tx).toMatchSnapshot()
    })
});


