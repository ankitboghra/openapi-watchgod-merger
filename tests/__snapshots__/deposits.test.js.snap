// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`openapi and watchgod 1`] = `
Object {
  "_latestStatus": "deposited",
  "_openapiTxStatus": "deposited",
  "_txSource": "openapi_deposits",
  "_watchgodTxStatus": "confirmed",
  "amount": "100000000000000000000",
  "amounts": null,
  "blockNumber": null,
  "from": "0xdepositer_1",
  "id": "id_1",
  "isPoS": null,
  "rootToken": "0xrootToken_1",
  "rootTunnelAddress": "0xrootTunnelAddress_1",
  "timestamp": 1639662463,
  "tokenId": null,
  "tokenIds": null,
  "tokenType": "tokenType_1",
  "txHash": "0xtxHash_1",
  "txStatus": "deposited",
  "userAddress": "0xuserAddress_1",
}
`;

exports[`openapi and watchgod 2`] = `
Object {
  "_latestStatus": "deposited",
  "_openapiTxStatus": "deposited",
  "_txSource": "openapi_deposits",
  "_watchgodTxStatus": "confirmed",
  "amount": "100000000000000000000",
  "amounts": null,
  "blockNumber": null,
  "from": "0xdepositer_1",
  "id": "id_1",
  "isPoS": null,
  "rootToken": "0xrootToken_1",
  "rootTunnelAddress": "0xrootTunnelAddress_1",
  "timestamp": 1639662463,
  "tokenId": null,
  "tokenIds": null,
  "tokenType": "tokenType_1",
  "txHash": "0xtxHash_1",
  "txStatus": "deposited",
  "userAddress": "0xuserAddress_1",
}
`;

exports[`openapi and watchgod 3`] = `
Object {
  "_latestStatus": "deposited",
  "_openapiTxStatus": "deposited",
  "_txSource": "openapi_deposits",
  "_watchgodTxStatus": "confirmed",
  "amount": "100000000000000000000",
  "amounts": null,
  "blockNumber": null,
  "from": "0xdepositer_2",
  "id": "id_2",
  "isPoS": null,
  "rootToken": "0xrootToken_2",
  "rootTunnelAddress": "0xrootTunnelAddress_2",
  "timestamp": 1639662463,
  "tokenId": null,
  "tokenIds": null,
  "tokenType": "tokenType_2",
  "txHash": "0xtxHash_2",
  "txStatus": "deposited",
  "userAddress": "0xuserAddress_2",
}
`;

exports[`openapi deposit only 1`] = `
Object {
  "_latestStatus": "deposited",
  "_openapiTxStatus": "deposited",
  "_txSource": "openapi_deposits",
  "amount": "100000000000000000000",
  "amounts": null,
  "blockNumber": null,
  "from": "0xdepositer_1",
  "id": "id_1",
  "isPoS": null,
  "rootToken": "0xrootToken_1",
  "rootTunnelAddress": "0xrootTunnelAddress_1",
  "timestamp": 1639662463,
  "tokenId": null,
  "tokenIds": null,
  "tokenType": "tokenType_1",
  "txHash": "0xtxHash_1",
  "txStatus": "deposited",
  "userAddress": "0xuserAddress_1",
}
`;

exports[`openapi deposits only 1`] = `
Object {
  "_latestStatus": "deposited",
  "_openapiTxStatus": "deposited",
  "_txSource": "openapi_deposits",
  "amount": "100000000000000000000",
  "amounts": null,
  "blockNumber": null,
  "from": "0xdepositer_1",
  "id": "id_1",
  "isPoS": null,
  "rootToken": "0xrootToken_1",
  "rootTunnelAddress": "0xrootTunnelAddress_1",
  "timestamp": 1639662463,
  "tokenId": null,
  "tokenIds": null,
  "tokenType": "tokenType_1",
  "txHash": "0xtxHash_1",
  "txStatus": "deposited",
  "userAddress": "0xuserAddress_1",
}
`;

exports[`openapi deposits only 2`] = `
Object {
  "_latestStatus": "deposited",
  "_openapiTxStatus": "deposited",
  "_txSource": "openapi_deposits",
  "amount": "100000000000000000000",
  "amounts": null,
  "blockNumber": null,
  "from": "0xdepositer_2",
  "id": "id_2",
  "isPoS": null,
  "rootToken": "0xrootToken_2",
  "rootTunnelAddress": "0xrootTunnelAddress_2",
  "timestamp": 1639662463,
  "tokenId": null,
  "tokenIds": null,
  "tokenType": "tokenType_2",
  "txHash": "0xtxHash_2",
  "txStatus": "deposited",
  "userAddress": "0xuserAddress_2",
}
`;

exports[`watchgod deposit only 1`] = `
Object {
  "_latestStatus": "confirmed",
  "_txSource": "watchgod_deposits",
  "_watchgodTxStatus": "confirmed",
  "amount": "100000000000000000000",
  "blockNumber": 13816531,
  "data": "0xdata_1",
  "from": "0xfrom_1",
  "network": "main",
  "rootToken": "0xrootToken_1",
  "timestamp": 1639662471018,
  "to": "0xto_1",
  "txBurnHash": null,
  "txHash": "0xtxHash_1",
  "txStatus": "confirmed",
  "txType": "deposit",
}
`;

exports[`watchgod deposits only 1`] = `
Object {
  "_latestStatus": "confirmed",
  "_txSource": "watchgod_deposits",
  "_watchgodTxStatus": "confirmed",
  "amount": "100000000000000000000",
  "blockNumber": 13816531,
  "data": "0xdata_1",
  "from": "0xfrom_1",
  "network": "main",
  "rootToken": "0xrootToken_1",
  "timestamp": 1639662471018,
  "to": "0xto_1",
  "txBurnHash": null,
  "txHash": "0xtxHash_1",
  "txStatus": "confirmed",
  "txType": "deposit",
}
`;

exports[`watchgod deposits only 2`] = `
Object {
  "_latestStatus": "confirmed",
  "_txSource": "watchgod_deposits",
  "_watchgodTxStatus": "confirmed",
  "amount": "100000000000000000000",
  "blockNumber": 13816531,
  "data": "0xdata_2",
  "from": "0xfrom_2",
  "network": "main",
  "rootToken": "0xrootToken_2",
  "timestamp": 1639662471018,
  "to": "0xto_2",
  "txBurnHash": null,
  "txHash": "0xtxHash_2",
  "txStatus": "confirmed",
  "txType": "deposit",
}
`;
